rfcs/20231017-collective-broadcast.md

```diff
@@ -0,0 +1,116 @@
+
+## Why call it collective_broadcast and not just broadcast?
+Unfortunately, the op name `broadcast` is already taken by [an op in XLA proper](https://www.tensorflow.org/xla/operation_semantics#broadcast), so we can't have the two names clash. `collective_broadcast` was the preferred alternative.
+
+## Proposed Specification
+
+### collective_broadcast
+
+#### Semantics
+
+Within each process group in the StableHLO process grid, send the value of the
+`operand` tensor from the source process to the target processes and produce a
+`result` tensor. The first process in each of the `replica_groups` is treated as the 
+source process. For any process `n` not in any replica group, that process will just return 
+zeros.
+
+The operation splits the StableHLO process grid into `process_groups` which is
+defined as follows:
+
+* `cross_replica(replica_groups)`
+  if `channel_id <= 0 and use_global_device_ids = false`.
+* `cross_replica_and_partition(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = false`.
+* `flattened_ids(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = true`.
+
+Afterwards, within each `process_group`:
+
+* `result@process[result_index] = operand@process[0]`
+
+#### Inputs
+
+| Label | Name                    | Type                                                             | Constraints |
+|-------|-------------------------|------------------------------------------------------------------|-------------|
+| (I1)  | `operand`               | tensor                                                           | (C5)        |
+| (I2)  | `replica_groups`        | variadic number of 1-dimensional tensor constants of type `si64` | (C1-C3)     |
+| (I3)  | `channel_id`            | constant of type `si64`                                          | (C4)        |
+| (I4)  | `use_global_device_ids` | constant of type `i1`                                            | (C4)        |
+
+#### Outputs
+
+| Name     | Type   | Constraints |
+|----------|--------|-------------|
+| `result` | tensor | (C5)        |
+
+#### Constraints
+
+* (C1) `is_unique(replica_groups)`.
```

Comment:
Most of these constraints will be outdated. Here's a new set of constraints for the updated semantics:
```
* (C1) is_unique(replica_groups[:, 0]).
* (C2) is_unique(replica_groups[:, 1:]).
* (C3) `0 <= replica_groups < N` where `N` is defined as:
  * `num_replicas` if `cross_replica` is used.
  * `num_partitions` if `cross_partition` is used.
* (C4) `type(result) = type(operand)`.
```

Reply:
I see, I was thinking that `[2,2]`, which is a no-op, is okay to have, but it would be good to just not allow this in the first place and simplify the constraints.